File: 48-rotate-image/rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int l = 0, r = n - 1;
        
        while (l < r) {
            for (int i = 0; i < r - l; i++) {
                int top,bottom;
                top = l;
                bottom = r;
                int topElement = matrix[top][l + i]; // save top
                matrix[top][l + i] = matrix[bottom - i][l]; // left to top
                matrix[bottom - i][l] = matrix[bottom][r - i]; // bottom to left
                matrix[bottom][r - i] = matrix[top + i][r]; // right to bottom
                matrix[top + i][r] = topElement; // top to right
            }
            l++;
            r--;
        }
    }
};

File: 48-rotate-image/rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        l ,r = 0 , n-1
        
        while(l < r):
            for i in range(0,r-l):
                top = l
                bottom = r
                topElement = matrix[top][l+i]
                matrix[top][l+i] = matrix[bottom-i][l]
                matrix[bottom-i][l] = matrix[bottom][r-i]
                matrix[bottom][r-i] = matrix[top + i][r]
                matrix[top+i][r] = topElement
            l += 1
            r -= 1
        